### PR TITLE
Scope pagination

### PR DESCRIPTION
**Pull request by johanbrandhorst**
Add pagination of scopes

### CODE REVIEW

File: internal/daemon/controller/handlers/scopes/scope_service_test.go [hunk 1]
```diff
@@ -575,10 +637,337 @@ func TestList(t *testing.T) {
...
+	_, err = sqlDb.ExecContext(ctx, "analyze")
+	require.NoError(t, err)
+
+	// request updated results
+	// since both creating and deleting scopes will affect the grantsHash
+	// we expect this to error
+	req.ListToken = got.ListToken
+	_, err = s.ListScopes(ctx, req)
+	require.Error(t, err)
+	require.True(t, errors.Match(errors.T(errors.InvalidListToken), err))
```
Comment (johanbrandhorst) [inline internal/daemon/controller/handlers/scopes/scope_service_test.go:876]: I couldn't get this not to error so I'm leaving it as it was in the previous PR.

File: internal/daemon/controller/handlers/scopes/scope_service_test.go [hunk 2]
```diff
@@ -575,10 +637,337 @@ func TestList(t *testing.T) {
...
+	requestInfo := authpb.RequestInfo{
+		TokenFormat: uint32(auth.AuthTokenTypeBearer),
+		PublicId:    at.GetPublicId(),
+		Token:       at.GetToken(),
+	}
+	requestContext := context.WithValue(context.Background(), requests.ContextRequestInformationKey, &requests.RequestContext{})
+	ctx = auth.NewVerifierContext(requestContext, iamRepoFn, tokenRepoFn, serversRepoFn, kms, &requestInfo)
+
+	req := &pbs.ListScopesRequest{
+		ScopeId:   oWithProjects.GetPublicId(),
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/daemon/controller/handlers/scopes/scope_service_test.go:742]: I think a recursive test would only test that the auth package returns the appropriate scopes from the Verify function and wouldn't provide much in the way of confidence that pagination is implemented correctly. We have existing list tests that test recursive listing, so I don't think it needs to be added here.

File: internal/daemon/controller/handlers/scopes/scope_service_test.go [hunk 3]
```diff
@@ -575,10 +637,337 @@ func TestList(t *testing.T) {
...
+		ListToken: "",
+		PageSize:  2,
+	}
+
+	s, err := scopes.NewService(ctx, repoFn, kms, 1000)
+	require.NoError(t, err)
+
+	// Run analyze in the DB to update the estimate tables
+	_, err = sqlDb.ExecContext(ctx, "analyze")
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/daemon/controller/handlers/scopes/scope_service_test.go]: Nope, good call! I've removed it.

File: internal/daemon/controller/handlers/scopes/scope_service_test.go [hunk 4]
```diff
@@ -5,8 +5,9 @@ package scopes_test
 
 import (
 	"context"
-	"errors"
+	stderrors "errors"
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/daemon/controller/handlers/scopes/scope_service_test.go]: Good shout, done!

**Review by mikemountain — APPROVED** [review]

**Pull request merged**